usr/src/uts/common/fs/zfs/zio.c

```diff
@@ -1417,6 +1417,13 @@ zio_taskq_dispatch(zio_t *zio, zio_taskq_type_t q, boolean_t cutinline)
 	spa_t *spa = zio->io_spa;
 	zio_type_t t = zio->io_type;
 	int flags = (cutinline ? TQ_FRONT : 0);
+	zio_priority_t p = zio->io_priority;
+
+	ASSERT(q == ZIO_TASKQ_ISSUE || q == ZIO_TASKQ_INTERRUPT);
+	if (p == ZIO_PRIORITY_SYNC_WRITE)
+		ASSERT3U(t, ==, ZIO_TYPE_WRITE);
```

Comment:
FYI, another way to do this is: `IMPLY(p == SYNC_WRITE, t == TYPE_WRITE)`

Reply:
Than you for the suggestion.  I decided to modify the code to use the more compact form.